### PR TITLE
Review Web Folder

### PR DESCRIPTION
**Pull request by mcandru**
Excuse the mess of branches. I had to do this to get the PR code review thing to work :)

### CODE REVIEW

File: Web/www/index.html [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+						<br>
+						<label style="display:inline-block; width:10%;" class="text--medium">D</label>
+						<input type="text">
+						<button class="button_up text--bold" type="increment">∧</button>
+						<button class="button_down text--bold" type="increment">∨</button>
+
+						<br><br>
+						<div style="text-align: center;">
+							<button class="text--bold" type="submit">Submit</button>
+						</div> -->
```
Comment (mcandru) [inline Web/www/index.html]: You can remove all of that commented shtuff to clean it up

File: Web/www/index.html [hunk 2]
```diff
@@ -0,0 +1,194 @@
...
+					</div>
+				</div>
+			</section>
+		</main>
+		<script>
```
Comment (mcandru) [inline Web/www/index.html]:
```suggestion
		<script type="text/javascript" src="path-to-javascript-file.js"></script>
```
My first suggestion here is to move all of the javascript to another file. Keeps everything clean. Oh and use 2 spaces rather than 4 with the javascript files. Looks much nicer.

File: Web/www/index.html [hunk 3]
```diff
@@ -0,0 +1,194 @@
...
+			var pid_card = document.getElementById("pid_card");
```
Comment (mcandru) [inline Web/www/index.html]:
```suggestion
			const pid_card = document.getElementById("pid_card");
```

Nowadays they have a version of Javascript called ES6. It has a whole bunch of nice improvements on Javascript of old.

One improvement that they made was better variable scoping. You can now declare variables as `const` or `let` which are much better variable definitions. [Here's](https://medium.com/javascript-scene/javascript-es6-var-let-or-const-ba58b8dcde75#:~:text=%60const%60%20is%20a%20signal%20that,always%20the%20entire%20containing%20function.) a crappy medium article that explains the difference.

File: Web/www/index.html [hunk 4]
```diff
@@ -0,0 +1,194 @@
...
+						console.log(document.getElementById(names[i] + "but_p_up").value);
+					});*/
+					// Input listeners
+					
+				}
+			}
+		</script>
+	</body>
+
+</html>
```
Comment (mcandru) [inline Web/www/index.html]:
```suggestion
</html>

```
Add a new line at the end of each file. [Here's](https://unix.stackexchange.com/questions/18743/whats-the-point-in-adding-a-new-line-to-the-end-of-a-file) a thread on some of the reasons to do it. It's pretty boring shite!

File: Web/www/index.html [hunk 5]
```diff
@@ -0,0 +1,194 @@
...
+				submit.id = "submit_btn";
+				var submit_contain = document.createElement("div");
+				submit_contain.style = "text-align: center;";
+				submit_contain.appendChild(submit);
+				pid_card.appendChild(_br.cloneNode(true));
+				pid_card.appendChild(submit_contain);
+			}
+			window.onload = function(){
+				// Submit button listener
+				// Should read all of the text boxes and save to pid.json
```
Comment (mcandru) [inline Web/www/index.html]: Is there any reason that you're saving to `pid.json`? How does this work exactly? Each time you start up the glider it reads from `pid.json`? I can't seem to find the file to be referenced anywhere else in the codebase.

File: Web/www/index.html [hunk 6]
```diff
@@ -0,0 +1,194 @@
...
+			var pid_card = document.getElementById("pid_card");
+			var names = ["Pitch", "Roll", "Yaw"];
+			fetch('pid.json')
+				.then(response => response.json())
+				.then(data => appendPID(data));
+			
+			function appendPID(data){
+				console.log(data);
+				// Run for pitch, roll and yaw
+				for (var i = 0; i < data.length; i++) {
```
Comment (mcandru) [inline Web/www/index.html]:
```javascript
data.forEach((property) => {
  // Put code in here
})
```
This is a really nice functional way of iterating over objects in ES6 javascript. Much cleaner than the `i++` of old.

File: Web/www/test.pyhtml [hunk 7]
```diff
@@ -0,0 +1,34 @@
+<html>
```
Comment (mcandru) [inline Web/www/test.pyhtml:1]: Have you gotten this `.pyhtml` file to work? Using one of these to look over the pid settings would probably be the cleanest way of doing it to be honest.

File: Web/www/pid.json [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+        "i": "0",
+        "d": "1"
+    },
+
+    {
+        "p": "0",
+        "i": "1",
+        "d": "0"
+    }
+]
```
Comment (mcandru) [inline Web/www/pid.json]:
```suggestion
[
  {
    "name": "pitch",
    "data": {
      "p": 1,
      "i": 2,
      "d": 3
    }
  },
  {
    "name": "roll",
    "data": {
      "p": 1,
      "i": 0,
      "d": 1
    }
  },
  {
    "name": "yaw",
    "data": {
      "p": 0,
      "i": 1,
      "d": 0
    }
  }
]
```
Would it be possible to change the structure of the data to something like this so that it's easier to see what PID values are associated with what controller?

File: Web/www/index.html [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+					var _br = document.createElement("br");
```
Comment (mcandru) [inline Web/www/index.html]:
```javascript
const propertyMarkup = `
  <div class="property-container">
    <p>${property.name}</p>
    ...
  </div>
`;
```
This is a decent way to do it if you can't use that `pyhtml` templating language thing. I'll give more details on that if you can't use the `.pyhtml` but I would say that it is much cleaner to do it that way.

**Review by mcandru — COMMENTED** [review]
Here's some initial comments on an approach to do it. It's more questions on how it's going to work more than anything else 😂 It looks super slick so far, love the clear design 👌

File: Web/www/index.html (hunk 2 above)
Comment (KevMca) [reply Web/www/index.html]: I thought as much 👍 I was putting off moving it to another file XD

File: Web/www/index.html (hunk 4 above)
Comment (KevMca) [reply Web/www/index.html]: Fair enough, files are weird

File: Web/www/index.html (hunk 5 above)
Comment (KevMca) [reply Web/www/index.html]: This is probably quite confusing alright. I haven't implemented the part to read and use the PID parameters anywhere else, but this is the thinking behind it...

I want to keep the PID values in non-volatile memory so that they are maintained through power cycles, but I also want programs to be able to modify them. Traditionally in embedded stuff, the program sits in flash and any non-volatile variables sit on an EEPROM (generally the flash is not used at runtime, just to store the program). Since the ESP has no EEPROM, a virtual EEPROM can be made if the arduino IDE is being used to program the device (using EEPROM.h).

However, the ESP32 is running micropython :) Therefore the microcontroller has a filesystem built in which allows flash memory to be accessed during runtime (which is pretty cool for a microcontroller). The server runs asynchronously and reads/writes to the JSON file (pid.json) on the ESP. The main glider script that runs the PID controllers can also access this file and update the PID parameters based on the values inside.

I think that some pyhtml would work well actually! Maybe I could write directly to JSON file?

TLDR; The JSON file is acting as a non-volatile area to store the PID parameters. The file doesn't come up anywhere else because I haven't implemented the back-end stuff yet.

File: Web/www/test.pyhtml (hunk 7 above)
Comment (KevMca) [reply Web/www/test.pyhtml:1]: I have gotten it running before. I think I'll take a look into it and see what I can do with it 😄

File: Web/www/pid.json (hunk 8 above)
Comment (KevMca) [reply Web/www/pid.json]: Fair enough 👍  I didn't have a clear structure before because I was trying to do some looping stuff 😬

File: Web/www/index.html (hunk 5 above)
Comment (mcandru) [reply Web/www/index.html]: So just for a more high level understanding, when does the ESP 32 read `pid.json`? Just whenever it boots up?

File: Web/www/test.pyhtml (hunk 7 above)
Comment (mcandru) [reply Web/www/test.pyhtml:1]: Yeah you should be able to write a little python loop that generates a new div for each of the pitch, roll and yaw fairly easily and that'll solve it all for you.

Comment (mcandru) [reply Web/www/test.pyhtml:1]: It should have the ability to safely change the `pid.json` file too I believe

File: Web/www/index.html (hunk 5 above)
Comment (KevMca) [reply Web/www/index.html]: It will read from it on a regular basis, so about every second or so. That way you can change the PID parameters in roughly real time and see how the plane reacts.

File: Web/www/test.pyhtml (hunk 7 above)
Comment (KevMca) [reply Web/www/test.pyhtml:1]: I just looked into it there and I think I'll stay away from it. It's not the most well documented thing in the world 🤔  Let me see if I can get away without it first.

File: Web/www/index.html (hunk 5 above)
Comment (mcandru) [reply Web/www/index.html]: Ah cool cool yeah that makes sense. Do you have the name of the HTML template generator library that uses the `.pyhtml` files by chance actually? Can't seem to find it online and it looks pretty cool.

File: Web/www/test.pyhtml (hunk 7 above)
Comment (mcandru) [reply Web/www/test.pyhtml:1]: Ah sorry only after seeing this now. Very easy to go down the JavaScript approach if needs be anyway not a bother at all

File: Web/www/index.html (hunk 5 above)
Comment (KevMca) [reply Web/www/index.html]: That's why I didn't look too far into it. However, if you go to https://github.com/jczic/MicroWebSrv and then open microWebTemplate.py, that should give you an idea. MicroWebSrv is the server that runs on the ESP32 in micropython 😄

File: Web/www/index.html (hunk 6 above)
Comment (KevMca) [reply Web/www/index.html]: I have this implemented now if you want to take a look. Much cleaner ❤️

File: Web/www/index.js [hunk 10]
```diff
@@ -0,0 +1,114 @@
...
+    // -----------------------------------------------------------------------
+    // -- Title
+    const title = `<br><p class="text--bold"> ${element["name"]}</p><br>`;
+    const titleDiv = document.createElement("div");
+    titleDiv.innerHTML = title;
+    pid_card.appendChild(titleDiv);
+    
+    // -----------------------------------------------------------------------
+    // -- Input and increment buttons
+    for (var key in element["data"]) {
```
Comment (mcandru) [inline Web/www/index.js:38]:
```suggestion
    Object.keys(element.data).forEach((key) => {
      ...
    });
```
This is a more functional way of getting the keys for each element in the dictionary. One thing to note is that unlike in Python, dictionaries in Javascript are just objects so you can also use `element.data` rather than having to reference like a Python dictionary each time.

**Review by mcandru — COMMENTED** [review]
This is really nice. Super clean now with the loops and template literals for generating the HTML markup in javascript. I can't approve this PR because I created it but if I could I would 😂
